File: app/src/main/java/com/grsu/guideapp/adapters/SaveAdapter.java
package com.grsu.guideapp.adapters;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Bitmap.CompressFormat;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;

import com.grsu.guideapp.R;
import com.grsu.guideapp.database.Test;
import com.grsu.guideapp.project_settings.Settings;
import com.grsu.guideapp.utils.CryptoUtils;
import com.grsu.guideapp.utils.StreamUtils;
import com.grsu.guideapp.utils.extensions.BitmapKt;
import com.squareup.picasso.Picasso;

import java.io.File;
import java.io.FileOutputStream;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okio.BufferedSink;
import okio.Okio;
import timber.log.Timber;

public class SaveAdapter {

    public static void saveImage(String url) {
        File file = new File(Settings.CONTENT);
        FileOutputStream ostream = null;
        try {
            Bitmap bitmap = Picasso.get().load(url)/*.resize(200, 150)*/.get();
            if (!file.exists()) {
                file.mkdirs();
            }
            File file1 = new File(file, CryptoUtils.hash(url));
            file1.createNewFile();

            ostream = new FileOutputStream(file1);
            bitmap.compress(CompressFormat.PNG, 0, ostream);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            StreamUtils.closeStream(ostream);
        }
    }

    public static byte[] saveIcon(Test helper, String url) {

        byte[] bytes;
        Resources resources = helper.mContext.getResources();
        try {
            Bitmap bitmap = Picasso.get().load(url).resize(59, 75).get();
            bytes = BitmapKt.toByteArray(bitmap);
        } catch (Exception e) {
            Timber.e(e, "saveIcon: %s", e.getMessage());
            Drawable drawable = resources.getDrawable(R.drawable.noicon);
            Bitmap bitmap = ((BitmapDrawable) drawable).getBitmap();
            bytes = BitmapKt.toByteArray(bitmap);
        }
        return bytes;
    }

    public static void saveAudio(String url) {
        if (url == null || url.isEmpty()) {
            return;
        }

        File file = new File(Settings.CONTENT);
        if (!file.exists()) {
            file.mkdirs();
        }

        BufferedSink sink = null;
        try {
            Request request = new Request.Builder().url(url).build();
            Response response = new OkHttpClient().newCall(request).execute();

            if (response.isSuccessful() && response.body() != null) {
                File downloadedFile = new File(Settings.CONTENT, StringKt.toMD5(url));
                Timber.e("saveAudio: %s", StringKt.toMD5(url));
                sink = Okio.buffer(Okio.sink(downloadedFile));
                sink.writeAll(response.body().source());
            }

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            StreamUtils.closeStream(sink);
        }
    }
}

File: app/src/main/java/com/grsu/guideapp/adapters/SaveAdapter.kt
package com.grsu.guideapp.adapters

import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import androidx.annotation.VisibleForTesting
import com.grsu.guideapp.App
import com.grsu.guideapp.R
import com.grsu.guideapp.utils.extensions.createIfNotExists
import com.grsu.guideapp.utils.extensions.toByteArray
import com.grsu.guideapp.utils.extensions.toMD5
import com.squareup.picasso.Picasso
import okhttp3.OkHttpClient
import okhttp3.Request
import okio.Okio
import java.io.File
import java.io.FileOutputStream

class SaveAdapter(private val client: OkHttpClient, basePath: File) {

    private val folder: File = File(basePath, "content")

    @VisibleForTesting
    fun downloadImageTest(url: String) {
        if (url.isEmpty()) return /*null*/

        val request = Request.Builder().url(url).build()
        val response = client.newCall(request).execute()
        response.body()?.also { body ->
            folder.createIfNotExists()
            val file = File(folder, url.toMD5() + ".png")
            val sink = Okio.buffer(Okio.sink(file))
            body.source().use { buffer -> sink.use { output -> output.writeAll(buffer) } }
        }
    }

    @VisibleForTesting
    fun downloadImageTest1(url: String, picasso: Picasso) {
        if (url.isEmpty()) return /*null*/
        val file = File(folder, url.toMD5())
        val bitmap: Bitmap = picasso.load(url).get()
        file.createIfNotExists()
        FileOutputStream(file).use { bitmap.compress(Bitmap.CompressFormat.PNG, 0, it) }
    }

    @VisibleForTesting
    fun saveMarker(url: String): ByteArray {
        return try {
            val bitmap = Picasso.get().load(url).resize(59, 75).get()
            bitmap.toByteArray()
        } catch (e: Exception) {
            val drawable = App.getInstance().getDrawable(R.drawable.noicon)
            (drawable as BitmapDrawable).bitmap.toByteArray()
        }
    }
}